### ISSUE TITLE
[Contributor Program Task 22] Profile and optimize text Hybrid-async on a low-resource Qwen3 setup

### ISSUE THREAD

**Issue by zheself** [redai-infra/Relax#124]
## Task

I would like to work on Contributor Program 2026 Cohort 1, Task 22:
**Hybrid-async text performance**.

The default entry is:

```text
scripts/training/text/run-qwen3-4B-8xgpu-hybrid-async.sh
```

I do not have stable access to eight GPUs, so I would like to confirm the
following reduced-resource entry before starting the formal benchmark.

## Repository and baseline

- Repository: `redai-infra/Relax`
- Baseline commit: `cfda05351d4b137f60e2b71d048beab1316d9d9f`
- Development branch: `perf/task-22-hybrid-async-text`
- Proposed recipe: `scripts/training/text/run-qwen3-0.6B-hybrid-async.sh`
- Delivery: this Issue plus a Draft PR with a guarded optimization, tests,
  recipe/documentation, and reproducible performance evidence

## Proposed environment

- Model: Qwen3-0.6B
- Training data: `zhuzilin/dapo-math-17k`
- Evaluation data: `zhuzilin/aime-2024`
- Algorithm: GRPO with the DAPO rule-based reward
- GPU: NVIDIA A40 48 GB
- Preliminary topology: 3 GPUs, Actor DP=2 plus one Rollout engine
- Primary benchmark topology: 4 GPUs, Actor DP=2 plus two Rollout engines
- Container: `ghcr.io/redai-infra/relaxrl:latest`
- Local SIF SHA-256:
  `dbd4c122f11e2e83f955ceeeadf541573c46f6458c47d892ce74c03794ed317e`
- Python/CUDA/framework versions: captured from the container in every formal
  run manifest before comparison

The 3-GPU topology is only for smoke testing and preliminary profiling. The
formal before/after claim will use the symmetric 4-GPU topology when four A40s
are available. Results from different GPU counts will never be used to compute
the speedup.

## Proposed baseline workload

```text
actor resources:                   [1 node, 2 GPUs]
rollout resources:                 [1 node, 2 GPUs]  # primary benchmark
tensor/pipeline/context parallel:  1 / 1 / 1
rollout batch size:                8 prompts
samples per prompt:                8
global batch size:                 64 samples
rollout max response length:       8192
rollout temperature:               1
max staleness:                     2
AIME interval:                     20
AIME samples per prompt:           8
AIME max response length:          16384
```

The recipe passes `--num-iters-per-train-update 1` for parity with the official
entry, but it is not treated as a Hybrid tuning axis. With
`8 * 8 / global_batch_size(64)`, the current Hybrid batch planner derives one
rollout mini-batch.

This preserves the effective DAPO workload of the official recipe while
reducing only the model size and hardware topology. In particular, the
reference forward path remains enabled even though `kl_loss_coef=0`, and no
sample count, response limit, or GRPO group will be removed to obtain a
performance gain.

Proposed training command inside the prepared Ray/Slurm container:

```bash
HYBRID_NUM_GPUS=4 \
MODEL_DIR=/path/to/models \
DATA_DIR=/path/to/data \
NUM_ROLLOUT=200 \
bash scripts/training/text/run-qwen3-0.6B-hybrid-async.sh
```

The complete Slurm allocation, container invocation, Ray address, CPU, memory,
image SHA-256, and package versions will be attached to the first baseline run
manifest before results are reported.

## Profiling scope and proposed optimization process

I will first separate:

1. Rollout generation and SGLang utilization;
2. Actor forward/backward/update time;
3. TransferQueue wait/backpressure;
4. Hybrid weight synchronization;
5. Pipeline bubbles between rollout and actor.

The final optimization will be selected from evidence rather than fixed in
advance. Candidate areas are resource balance, queue/staleness configuration,
dynamic batching, or a small guarded observability/scheduling change. If an
existing parameter combination is sufficient, the PR will prefer a recipe and
documentation change over a core Controller change.

Any core Controller, public argument, service, or launcher change will be
discussed before implementation.

## Metrics and statistical protocol

Baseline and optimized configurations will each be run at least twice on the
same node, GPU count/type, model, data, batch, sequence limits, sampling
settings, seed set, container, and measurement interval.

Each run will contain at least three predefined steady-state windows after
warm-up. I will report:

- end-to-end samples/s and, when available, generated tokens/s;
- step, rollout, actor-train, and weight-sync wall time (mean/median/p95);
- per-GPU average utilization and peak memory;
- queue wait/backpressure and actual staleness distribution;
- host peak memory when changing queue capacity or `max_staleness`;
- processed prompt/sample/token/update counts.

AIME evaluation can be much longer than a training step, especially with one
Rollout engine. Training-only steady-state windows will therefore exclude
periodic AIME intervals, while AIME duration/pass rate and an additional
evaluation-inclusive end-to-end wall time will be reported separately.

## Acceptance target and quality guardrails

Proposed primary target:

- improve mean steady-state samples/s by at least 5%, with the gain larger than
  run-to-run noise, or equivalently reduce steady-state step time without
  reducing effective work;
- no regression in the agreed quality/correctness guardrails.

Guardrails:

- exactly 64 samples and 8 complete GRPO groups per training step;
- no dropped samples, silent retries, NaN/Inf, OOM, or unexplained errors;
- actual staleness never exceeds the configured bound;
- reward/loss/KL/grad-norm and AIME pass-rate curves remain valid;
- response length, valid token count, and completion rate do not regress due to
  reduced work;
- the optimization has an explicit off switch and restores baseline behavior.

For `max_staleness=2`, the planned TransferQueue capacity is 192 samples
(`8 prompts * 8 samples * (2 + 1)`). Any staleness experiment will report the
corresponding capacity and host-memory effect.

## Scope boundaries

I will not:

- compare 3-GPU results directly against 4-GPU results as a speedup;
- claim that a Qwen3-0.6B result automatically generalizes to Qwen3-4B/8-GPU;
- change model architecture, data, reward math, effective batch, sample count,
  response limits, or training work to manufacture throughput;
- include checkpoints, datasets, credentials, internal paths, or raw large logs
  in the PR.

Please confirm whether the Qwen3-0.6B + DAPO/AIME setup and the 3-GPU
preliminary / 4-GPU primary topology are acceptable for Task 22 before I start
the formal benchmark.

**Comment by MaybeIcanShow** [comment]
## 最终方案与实验结果更新

Task 22 的实现和三组正式配对实验已经完成，Draft PR：https://github.com/redai-infra/Relax/pull/211

### 与最初提案的差异

受最终可用硬件和可重复运行窗口影响，正式实验固定为：

- 上游基线：`redai-infra/Relax main@0bc99af`；
- 实验代码提交：`056f4b47c2022910c780f497d42283a803dc2ea7`；
- 开发分支：`MaybeIcanShow:perf/task22-hybrid-async-zero-kl`；
- 模型：Qwen3-0.6B；数据：ModelScope `AI-ModelScope/gsm8k` main 固定 16 prompts；
- 硬件：2× NVIDIA RTX PRO 6000 Blackwell，Actor/Rollout 各 1 GPU；
- 每个配置 3 次运行，每次 20 steps；稳定窗口为记录 step 5-15；
- 每 step 8 prompts × 4 samples，有效 batch 32，response cap 512。

因此，本结果只证明上述 2 GPU 小模型配置下的性能收益，不外推到最初计划的 4×A40 或官方 Qwen3-4B/8 GPU recipe。

### 最终优化

1. `zero_kl`：KL loss 系数严格为 0 时，不加载 reference checkpoint、不执行 reference forward；
2. `optimized`：在 `zero_kl` 基础上设置 `--update-weights-interval 2`，每两次 Actor update 发布一次 Rollout 权重；
3. 评测边界、epoch 边界和最后一步强制发布；默认 interval=1 保持兼容；
4. `TASK22_VARIANT=zero_kl UPDATE_WEIGHTS_INTERVAL=1` 可回退间隔优化，`TASK22_VARIANT=baseline` 可回退全部改动。

### 优化动机与理论解释

Hybrid-async 同时推进 Rollout 生成和 Actor 训练，但一次权重发布仍包含有同步成本的控制序列：暂停 Rollout、等待在途请求到达安全边界、传输 Actor 权重、结束更新并恢复生成。Qwen3-0.6B 单步训练较短时，这段固定成本占比更突出，逐 step 发布容易形成流水空泡。

将发布间隔从 1 调整为 2 后，两个 Actor update 共同摊销一次 pause/transfer/resume 成本。若单次发布耗时为 `T_sync`，逐步发布的平均开销约为 `T_sync`，interval-two 约为 `T_sync/2`；实际收益还取决于 Actor/Rollout 重叠、队列等待和生成长度，因此主结论来自端到端配对实验，而非只按同步耗时推算。

该优化会让 Rollout 最多额外使用一次 Actor update 之前的权重，属于以 policy freshness 换流水利用率。实验保持 `--max-staleness 2` 并启用 TIS，在评测、epoch 和最终 step 前强制发布；TIS、clip fraction、loss、reward、样本数和异常计数作为正确性护栏。

baseline 保留 `--use-kl-loss --kl-loss-coef 0.00` 和 reference checkpoint。KL 项严格乘以 0，因此移除 reference forward 不改变标量目标，只消除无效计算。设置 `zero_kl` 中间组后，`optimized / zero_kl` 才是 interval-two 的主验收对比。

### 效果数据

| 变体 | 响应 tok/s | samples/s | 整体 GPU 利用率 | Actor GPU 利用率 | Rollout GPU 利用率 | Actor 峰值 MiB |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| baseline | 4206.2 | 8.220 | 55.73% | 76.97% | 34.49% | 43978 |
| zero_kl | 4460.5 | 8.721 | 55.66% | 74.29% | 37.02% | 43974 |
| optimized | 5061.8 | 9.895 | 63.47% | 85.12% | 41.82% | 51168 |

- 移除零系数 KL reference 路径：响应吞吐 `+6.05%`；
- interval-two 相对 `zero_kl`：响应吞吐 `+13.48%`；
- optimized 相对 baseline：响应吞吐 `+20.34%`；
- 9 个作业均完成 20 steps，每个作业生成 640 samples；
- 意外 NaN/Inf、运行时错误、OOM 和样本丢失均为 0；
- optimized Actor 峰值显存增加 7190 MiB，作为性能收益的显存代价记录。

### 改动前后曲线

附件包包含三张三组配对实验稳定窗口曲线，均补齐横轴 `配对运行 / 训练 step`、纵轴单位、数值刻度、网格和运行分隔线：

- `throughput_curves.svg`：baseline、zero-KL 和 optimized 的逐 step 响应吞吐；
- `step_time_curves.svg`：三变体逐 step framework step 耗时；
- `weight_update_curves.svg`：三变体前序权重发布耗时，直接展示 interval-two 的开销变化。

报告、CSV、曲线和脱敏原始日志不提交仓库，统一作为 Draft PR 附件交付：

- `task22-hybrid-async-text.zip` SHA-256：`7b68b08a87429348590a94100277eb9c79c5140b31923ae3f89618efa160c694`；
- 内含 `raw-evidence.tar.gz` SHA-256：`8cf759a2fbed8baacf149985881d66aac71a4ab92066b523caf5790ee2568237`。

附件待通过 GitHub PR 网页上传；上传后将在此处和 PR 正文补充 `user-attachments` 长期链接。

### 正确性边界

这是短周期性能实验。大部分 step 的 DAPO reward 没有组内方差，不能据此证明 interval-two 的长期收敛等价。PR 保持 Draft，并明确请求维护者确认是否需要更长、持续产生非零梯度的质量对比。

### 复现与验证

```bash
MODEL_PATH=/path/to/Qwen3-0.6B CUDA_VISIBLE_DEVICES=2,3 TOTAL_TRIALS=3 \
  bash benchmarks/task22_hybrid_async_text/run_paired_trials.sh
```

Python 3.12 下 `pre-commit run --all-files` 已完整通过；Task 22 定向测试为 `16 passed, 1 skipped`，跳过项缺少可选 Megatron 依赖；原始证据包 SHA-256 校验通过。

**Comment by MaybeIcanShow** [comment]
## Task 22 v4 实验结果更新

PR #211 的评审修复已在提交 `a0b16761cd8df2828cba08e05b5c861fef7924da` 完成，并在实验条件不变的情况下串行重跑了 3 变体 × 3 次配对实验。

固定条件：Qwen3-0.6B、2× NVIDIA RTX PRO 6000 Blackwell（Actor/Rollout 各 1 GPU）、GSM8K 固定 16 prompts、20 steps、step 5-15 稳定窗口、8 prompts × 4 samples（batch 32）、response cap 512、max staleness 2。

### 结果

| 变体 | 响应 tok/s | samples/s | framework/E2E step | 整体 GPU 利用率 |
| --- | ---: | ---: | ---: | ---: |
| baseline | 4449.5 | 8.701 | 3.678 / 3.697 s | 57.35% |
| zero_kl | 4484.9 | 8.767 | 3.650 / 3.667 s | 56.16% |
| optimized | 5071.9 | 9.910 | 3.229 / 3.273 s | 62.43% |

本次代码会在 `kl_loss_coef == 0` 时自动关闭 reference 路径，所以 `baseline` 和 `zero_kl` 是自动/显式配置等价性对照，吞吐差异仅 `+0.80%`。Task 22 的主验收对比是 interval-two `optimized` 相对 `zero_kl`：

- 平均响应吞吐 `+13.09%`；
- 三次配对分别 `+13.52% / +13.39% / +12.36%`，都超过 5% 目标；
- E2E 延迟 `-10.74%`；
- 每 step 权重发布耗时从 0.665 s 降至 0.304 s。

9/9 作业均完成 20 steps，每作业生成 640 samples；意外 NaN/Inf、runtime error、OOM 和样本丢失均为 0。所有变体都没有 reference forward，interval-one/interval-two 的发布次数分别为 21/11（含初始化发布）。

optimized Actor 峰值显存增加 7194 MiB，日志定位为跳过 step 未清理的可回收 PyTorch reserved allocator cache 高水位，不是 live tensor 或工作量增大。

### 证据

- [Task 22 v4 release assets](https://github.com/MaybeIcanShow/Relax/releases/tag/task22-evidence-v4)；
- [raw-evidence.tar.gz](https://github.com/MaybeIcanShow/Relax/releases/download/task22-evidence-v4/raw-evidence.tar.gz) SHA-256：`da497ee37d05f97e43c5c6277b21dd7793ee0c4236587d4fc60e054f78674ff6`；
- [raw-evidence.sha256](https://github.com/MaybeIcanShow/Relax/releases/download/task22-evidence-v4/raw-evidence.sha256)。

release 以 9 个独立 asset 发布完整报告、summary/逐 step CSV、三张曲线、脱敏原始日志包和文件级 SHA-256 索引。`pre-commit` 通过，Task 22 定向测试为 `29 passed, 1 skipped`，跳过项缺少可选 Megatron 依赖。

结论仅适用于上述 2 GPU 短周期性能实验，不外推到 Qwen3-4B/8 GPU，也不证明长期收敛等价。

**Comment by MaybeIcanShow** [comment]
## PR #211 最新修复与 v6 复测结果

Task22 PR 已根据最新评审继续修复并重新提交：

- 核心修复：`946a0dc6`；benchmark 证据工具修复：`6d2d71b`；
- global dataset epoch 长度在所有 service 创建前注入；
- non-global dataset 启动与 eval interval 路径已打通；
- pause 不确定、recover 失败、resume 失败及权重更新异常均有补偿和 all-rank 一致失败处理；
- HTTP health/recover/evaluate/resume 请求补齐 timeout；
- Task22 定向测试 `42 passed`，pre-commit 通过。

固定 Qwen3-0.6B / 2 GPU / GSM8K 16 prompts 配置下，最终 9/9 作业均完成 20 steps：

| 变体 | 响应 tok/s | samples/s | 发布耗时/step |
| --- | ---: | ---: | ---: |
| baseline | 4465.7 | 8.727 | 0.673 s |
| zero_kl | 4431.7 | 8.668 | 0.680 s |
| optimized | 5039.1 | 9.851 | 0.318 s |

`optimized` 相对 `zero_kl` 提升 `+13.71%`，三次配对分别为 `+14.10% / +13.00% / +14.03%`，均超过 5% 验收目标；无意外 NaN/Inf、runtime error、OOM 或样本丢失。

完整证据：[Task 22 v6 release](https://github.com/MaybeIcanShow/Relax/releases/tag/task22-evidence-v6)

`raw-evidence.tar.gz` SHA-256：`740f39d14c993d9240d646ddb6600ceff9c17c12810689ba765134f6c8efe2cf`。

**Comment by MaybeIcanShow** [comment]
## PR #211 代码规范与交付方式更新

最新提交：`de5fd69e6f14b4618cc8916dace54864670e463a`。

本轮按代码规范和交付范围完成以下调整：

- PR 已移除 `benchmarks/` 下的源码、固定数据，以及对应的 `tests/benchmarks/`；
- benchmark 源码、数据和测试改为通过个人 fork Release 交付，不进入上游 PR；
- 测试不再用宽泛的模块级 `try/except ImportError` 隐藏导入异常，改为对可选依赖使用显式 `pytest.importorskip()`；
- zero-KL 参数测试已从 OPD 专用测试文件拆分到独立测试模块，共用 fixture 移入 `conftest.py`；
- 两 GPU recipe 不再依赖仓库内 benchmark 数据准备脚本，要求调用方通过 `PROMPT_DATA` 指向准备好的 JSONL。

交付链接：

- [PR #211](https://github.com/redai-infra/Relax/pull/211)
- [Task 22 v6 Release](https://github.com/MaybeIcanShow/Relax/releases/tag/task22-evidence-v6)
- [benchmark 源码包](https://github.com/MaybeIcanShow/Relax/releases/download/task22-evidence-v6/task22-benchmark-source.tar.gz)
- 源码包 SHA-256：`3ab86aea9b12dbc9015efb52bec07d5b09bab3d77b43a7a3e568aff6733a47ef`

验证结果：定向测试 `16 passed, 3 skipped`（缺少可选 Megatron 依赖时显式 skip）；Ruff lint/format、`compileall`、shell 语法和 `git diff --check` 均通过。

GitHub 远端 PR 文件列表已确认不包含 `benchmarks/` 或 `tests/benchmarks/`。PR 当前仍为 Draft，并与最新上游 `main` 存在合并冲突，需在转 Ready 前处理。
